### PR TITLE
[PPS] Timing calibration for diamond rechits/tracks

### PR DESCRIPTION
**Pull request by forthommel**
#### PR description:

This PR modifies the rechits producer for PPS diamond detectors to insert the calibrated hit time of arrival, and propagates this information onto a general track time, following the same recipe as introduced #25465 for TOTEM vertical timing detectors.
As most of the algorithmic part is common, the overall track-from-hits timing extraction method is moved to the base class.
The calibration objects are also shared between the two reconstruction codes, introduced in #25777 (with ESSource modified in #26207 for diamond calibration files parsing).
This track timing information is propagated (with no code change) up to the `CTPPSLocalTrackLite` miniAOD-optimised object.

#### PR validation:

`runTheMatrix.py -l limited` test ongoing.

**Warning** as no GT is currently holding the diamond timing calibration for all periods defined in 2017-18, it its current state, this PR should return an uncorrected (thus unmodified) rechits collection.

#### if this PR is a backport please specify the original PR:

Before submitting your pull requests, make sure you followed this checklist:
- [x] verify that the PR is really intended for the chosen branch
- [x] verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- [x] verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by forthommel** [comment]
Beside the removal of a `unsigned int` → `int` cast for the digi leading/trailing edge time  ([this](https://github.com/cms-sw/cmssw/pull/26327/files#diff-f2eceb8f62d597e0f4538adf55a3321eL47) to [that](https://github.com/cms-sw/cmssw/pull/26327/files#diff-f2eceb8f62d597e0f4538adf55a3321eR58)) I do not spot any algorithmic difference that might lead to this difference in the time over threshold distribution for 135.85.
image
Beside this, is there a CMS guideline for the `std::sqrt`/`sqrt` convention (equivalently for `pow`) leading to this [header consistency check failure](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26327/33971/headers_chks.log)?

**Comment by slava77** [comment]
@forthommel 
for this and some other PPS PRs, may I ask you to add a link to a slide/table in the PR description with a summary of data taking periods where different PPS detectors were present.
This should help to better interpret presence or absence of changes in the tested workflows.
Thank you.

**Comment by jan-kaspar** [comment]
@slava77 I could suggest for example Valentina's slides 3 to 6 here: 
https://indico.cern.ch/event/791284/contributions/3305099/attachments/1809481/2954649/RUN-DPG_workshop_PPS.pdf

**Comment by slava77** [comment]
> @slava77 I could suggest for example Valentina's slides 3 to 6 here:
> https://indico.cern.ch/event/791284/contributions/3305099/attachments/1809481/2954649/RUN-DPG_workshop_PPS.pdf

The granularity in these slides is one year, but IIUC, the actual periods were smaller than the full year at least for some detectors.

**Comment by jan-kaspar** [comment]
> > @slava77 I could suggest for example Valentina's slides 3 to 6 here:
> > https://indico.cern.ch/event/791284/contributions/3305099/attachments/1809481/2954649/RUN-DPG_workshop_PPS.pdf
> 
> The granularity in these slides is one year, but IIUC, the actual periods were smaller than the full year at least for some detectors.

The "sensor types" were changed with year granularity as summarised in the presentation. Then, for some runs/periods some RPs may not have delivered data (failures, RPs not inserted, ...). For 2017 and 2018, the status of each RP can be retrieved from run registry. For 2016 (validation not yet done), if you have a specific run in mind, we can tell what the RP status was.

File: RecoCTPPS/TotemRPLocal/interface/CTPPSTimingTrackRecognition.h [hunk 1]
```diff
@@ -191,34 +200,49 @@ CTPPSTimingTrackRecognition<TRACK_TYPE, HIT_TYPE>::getHitSpatialRange(const HitV
...
+template<class TRACK_TYPE, class HIT_TYPE> inline
+bool
+CTPPSTimingTrackRecognition<TRACK_TYPE, HIT_TYPE>::timeEval(const HitVector& hits, float& mean_time, float& time_sigma) const
+{
+  float mean_num = 0.f, mean_denom = 0.f;
+  bool valid_hits = false;
+  for (const auto& hit : hits) {
+    if (hit.getTPrecision() <= 0.)
+      continue;
+    const float weight = std::pow(hit.getTPrecision(), -2);
```
Comment (slava77) [inline RecoCTPPS/TotemRPLocal/interface/CTPPSTimingTrackRecognition.h]: header consistency check flags the use of ```pow``` as problematic.
Please include ```<cmath>```

Comment (forthommel) [reply RecoCTPPS/TotemRPLocal/interface/CTPPSTimingTrackRecognition.h]: Resolved in https://github.com/cms-sw/cmssw/pull/26327/commits/9e32e5ffde85854a01410ebb0c180073996fdb5d

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoCTPPS/TotemRPLocal/src/CTPPSDiamondRecHitProducerAlgorithm.cc [hunk 2]
```diff
@@ -34,25 +45,49 @@ CTPPSDiamondRecHitProducerAlgorithm::build( const CTPPSGeometry* geom, const edm
...
+      if ( t_lead == 0 && t_trail == 0 )
+        continue;
+
+      double tot = -1., ch_t_twc = 0.;
+      if ( t_lead != 0 && t_trail != 0 ) {
+        tot = ( t_trail-t_lead )*ts_to_ns_; // in ns
+        if ( calib_fct_ ) {
+          // compute the time-walk correction
+          ch_t_twc = calib_fct_->evaluate( std::vector<double>{ tot }, ch_params );
+          if ( std::isnan( ch_t_twc ) )
```
Comment (slava77) [inline RecoCTPPS/TotemRPLocal/src/CTPPSDiamondRecHitProducerAlgorithm.cc]: static analyzer flagged this
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26327/34130/llvm-analysis/report-2166ef.html#EndPath
```Please use edm::isNotFinite from 'FWCore/Utilities/interface/isNotFinite.h'```

Comment (forthommel) [reply RecoCTPPS/TotemRPLocal/src/CTPPSDiamondRecHitProducerAlgorithm.cc]: Good catch! However the include path seems outdated... Attempted to fix in #26435.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by pohsun** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/26327	149e399
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show only a difference in ctppsDiamondRecHits time over threshold (ToT) as noted already in https://github.com/cms-sw/cmssw/pull/26327#issuecomment-480198473 as expected

**Comment by fabiocos** [comment]
+1

**Pull request merged**